### PR TITLE
[GPU] Weights reorders primitive cachcing

### PR DESCRIPTION
**Pull request by sshlyapn**
### Details:
 - Moved generic_layer primitive to intel_gpu/primitives/
 - Added WeightsReorderParams and WeightsReorderParamsOCL for weights reorders description
 - Replaced in_mem_kernels_cache with common kernels_cache
 - Removed all CPU related generic_layer's options

### Tickets:
 - 106743

### CODE REVIEW

File: src/plugins/intel_gpu/src/graph/impls/ocl/primitive_base.hpp [hunk 1]
```diff
@@ -207,6 +205,18 @@ struct typed_primitive_impl_ocl : public typed_primitive_impl<PType> {
...
 
+    void set_arguments_impl(typed_primitive_inst<PType>& instance, kernel_arguments_data& args) override {
+        if (instance.can_be_optimized() || is_cpu()) {
+            return;
+        }
+
+        stream& stream = instance.get_network().get_stream();
+
+        for (size_t k = 0; k < _kernels.size(); ++k) {
+            stream.set_arguments(*_kernels[k], _kernel_data.kernels[k].params, args);
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/graph/impls/ocl/primitive_base.hpp:216]: I think skipped kernels need to be handled here

File: src/plugins/intel_gpu/src/graph/impls/ocl/primitive_base.hpp [hunk 2]
```diff
@@ -207,6 +205,18 @@ struct typed_primitive_impl_ocl : public typed_primitive_impl<PType> {
         }
     }
 
+    void set_arguments_impl(typed_primitive_inst<PType>& instance, kernel_arguments_data& args) override {
+        if (instance.can_be_optimized() || is_cpu()) {
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/graph/impls/ocl/primitive_base.hpp:209]: This `is_cpu()` check seems to be useless, as classes derived from `typed_primitive_impl_ocl` are always non-cpu impls

File: src/plugins/intel_gpu/src/graph/include/implementation_map.hpp [hunk 3]
```diff
@@ -131,4 +131,28 @@ class implementation_map {
...
+            impl_types impl_type = kv.first.first;
+            shape_types supported_shape_type = kv.first.second;
+            if ((preferred_impl_type & impl_type) != impl_type)
+                continue;
+            if ((target_shape_type & supported_shape_type) != target_shape_type)
+                continue;
+
+            return kv.second;
+        }
+        OPENVINO_ASSERT(false, "[GPU] WeightsReordersFactory doesn't have any implementation for "
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/graph/include/implementation_map.hpp:154]: OPENVINO_THROW

File: src/plugins/intel_gpu/src/graph/impls/ocl/generic_layer.cpp [hunk 4]
```diff
@@ -27,18 +27,21 @@ struct generic_layer_impl : typed_primitive_impl<generic_layer> {
...
-        , _kernels()
-        , _cached_kernel_id() { }
+    generic_layer_impl(const kernel_impl_params& params)
+        : _cl_kernel_data()
+        , _kernel(nullptr)
+        , _cached_kernel_id() {
+        auto reorder_params = params.typed_desc<generic_layer>()->params;
+        auto casted_params = std::dynamic_pointer_cast<WeightsReorderParamsOCL>(reorder_params);
+        OPENVINO_ASSERT(casted_params, "[GPU] Invalid weights reorder parameters type for ", params.desc->id, " node");
+        _cl_kernel_data = *casted_params->get_cl_kernel();
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/impls/ocl/generic_layer.cpp:43]: Umm, are we going to define & confine generic layer only for weight reorder? 
If so.. why do we need "generic" layer?

File: src/plugins/intel_gpu/src/graph/include/primitive_inst.h [hunk 5]
```diff
@@ -94,7 +96,25 @@ struct primitive_impl {
     virtual void set_kernels(cldnn::kernels_cache::compiled_kernels kernels) {}
     virtual std::vector<kernel::ptr> get_kernels() { return {}; }
 
+    bool need_weights_reorder() const { return _weights_reorder_params != nullptr; }
+    std::shared_ptr<WeightsReorderParams> get_weights_reorder_params() const { return _weights_reorder_params; }
+    void reset_weights_reorder_params() { _weights_reorder_params = nullptr; }
+
+    std::shared_ptr<kernel_impl_params> get_weights_reorder_kernel_params() const {
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/include/primitive_inst.h:103]: Could we move the code body to primitive_inst.cpp to reduce future build time?

File: src/plugins/intel_gpu/src/graph/impls/ocl/generic_layer.cpp (hunk 4 above)
Comment (sshlyapn) [reply src/plugins/intel_gpu/src/graph/impls/ocl/generic_layer.cpp:43]: Yes, actually it is already used only for weights reordering, but maybe we need to consider to rename it for clarity...

File: src/plugins/intel_gpu/src/graph/include/primitive_inst.h (hunk 5 above)
Comment (sshlyapn) [reply src/plugins/intel_gpu/src/graph/include/primitive_inst.h]: Sure, moved to primitive_inst.cpp

**Review by vladimir-paramuzov — APPROVED** [review]

**Pull request merged**